src/photon/photon_algorithm.c

```diff
@@ -168,78 +182,144 @@ void fetch_missing_dependencies(local_scheduler_state *state,
+ * dependencies.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param spec The task specification to queue.
+ * @return Void.
+ */
+void queue_waiting_task(local_scheduler_state *state,
+                        scheduling_algorithm_state *algorithm_state,
+                        task_spec *spec) {
+  LOG_DEBUG("Queueing task in waiting queue");
+  /* Initiate fetch calls for any dependencies that are not present locally. */
+  if (plasma_manager_is_connected(state->plasma_conn)) {
+    fetch_missing_dependencies(state, algorithm_state, spec);
+  }
+  queue_task(state, &algorithm_state->waiting_task_queue, spec);
+}
+
+/**
+ * Queue a task whose dependencies are ready. When the task reaches the front
+ * of the dispatch queue and workers are available, it will be assigned.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param spec The task specification to queue.
+ * @return Void.
+ */
+void queue_dispatch_task(local_scheduler_state *state,
+                         scheduling_algorithm_state *algorithm_state,
+                         task_spec *spec) {
+  LOG_DEBUG("Queueing task in dispatch queue");
+  queue_task(state, &algorithm_state->dispatch_task_queue, spec);
+}
+
+/**
+ * Add the task to the proper local scheduler queue. This assumes that the
+ * scheduling decision to place the task on this node has already been made,
+ * whether locally or by the global scheduler.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param spec The task specification to queue.
+ * @return Void.
+ */
+void queue_task_locally(local_scheduler_state *state,
+                        scheduling_algorithm_state *algorithm_state,
+                        task_spec *spec) {
+  if (can_run(algorithm_state, spec)) {
+    /* Dependencies are ready, so push the task to the dispatch queue. */
+    queue_dispatch_task(state, algorithm_state, spec);
+  } else {
+    /* Dependencies are not ready, so push the task to the waiting queue. */
+    queue_waiting_task(state, algorithm_state, spec);
+  }
+}
+
+/**
+ * Give a task to the global scheduler to schedule.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param spec The task specification to schedule.
+ * @return Void.
+ */
 void give_task_to_global_scheduler(local_scheduler_state *state,
                                    scheduling_algorithm_state *algorithm_state,
-                                   task_spec *spec,
-                                   bool from_global_scheduler) {
+                                   task_spec *spec) {
+  if (state->db == NULL || !state->global_scheduler_exists) {
+    /* A global scheduler is not available, so queue the task locally. */
+    queue_task_locally(state, algorithm_state, spec);
+    return;
+  }
```

Comment:
No, it will get executed before the DCHECKs. This is for testing purposes, where there might not be a global scheduler.